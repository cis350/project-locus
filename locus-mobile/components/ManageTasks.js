/* eslint-disable object-curly-newline */
import React, { useState, useEffect } from 'react';
import {
  View, Text, StyleSheet, ScrollView, TouchableHighlight, Alert, TextInput, Modal,
} from 'react-native';
import { getAllTasksForProject, createTask } from '../modules/api';
import Task from './Task';

export default function ManageTasks({ project, setManagingTask, user, club }) {
  const [modalVisible, setModalVisible] = useState(false);
  const [jawn, rerender] = useState(false);
  const [tasks, setTasks] = useState(project.tasks);
  const [newTaskName, setNewTaskName] = useState('');
  const [newTaskAssignee, setNewTaskAssignee] = useState('');
  const [selectedTask, setSelectedTask] = useState(undefined);

  useEffect(() => {
    async function getTasks() {
      setTasks((await getAllTasksForProject(club.clubName, project.projectName, user.email))
        .jsonContent);
    }
    getTasks();
  }, [jawn, selectedTask]);

  async function handleCreateTask() {
    const response = await (
      createTask(
        club.clubName,
        project.projectName,
        newTaskName,
        user.email,
        newTaskAssignee,
        'incomplete',
      ));
    setNewTaskName('');
    setNewTaskAssignee('');
    setModalVisible(false);
    if (response.status === 201) Alert.alert('Task Successfully Created');
    else Alert.alert('Task Creation Failed');
    rerender(!jawn);
  }

  const displayTasks = [];
  for (let i = 0; i < tasks.length; i += 1) {
    displayTasks.push(
      <View style={styles.task} key={`task${i}`}>
        <Text style={styles.taskTitle}>{tasks[i].taskName}</Text>
        <Text style={styles.status}>{tasks[i].status}</Text>
        <TouchableHighlight style={styles.viewButton} onPress={() => setSelectedTask(tasks[i])} underlayColor="#33E86F">
          <Text style={{ textAlign: 'center', fontSize: 15, color: 'white' }}>View</Text>
        </TouchableHighlight>
      </View>,
    );
  }

  if (selectedTask) {
    return (
      <Task
        task={selectedTask}
        setSelectedTask={setSelectedTask}
        user={user}
        project={project}
        club={club}
      />
    );
  }
  return (
    <View style={styles.container}>
      <View style={styles.centeredView}>
        <Modal
          animationType="slide"
          transparent={false}
          visible={modalVisible}
        >
          <View style={styles.centeredView}>
            <Text style={{ fontSize: 40, marginBottom: 30 }}>Create Task</Text>
            <TextInput
              style={styles.input}
              placeholder="Task Name"
              placeholderTextColor="grey"
              onChangeText={setNewTaskName}
              value={newTaskName}
            />
            <TextInput
              style={styles.input}
              placeholder="Assignee Email"
              placeholderTextColor="grey"
              onChangeText={setNewTaskAssignee}
              value={newTaskAssignee}
            />
            <TouchableHighlight style={styles.button} underlayColor="#33E86F" onPress={() => handleCreateTask()}>
              <Text style={{ fontSize: 20 }}>Create</Text>
            </TouchableHighlight>
            <TouchableHighlight style={styles.cancelButton} underlayColor="#33E86F" onPress={() => setModalVisible(false)}>
              <Text style={{ fontSize: 20 }}>Cancel</Text>
            </TouchableHighlight>
          </View>
        </Modal>
      </View>
      <Text style={styles.projectTitle}>Tasks</Text>
      <ScrollView style={styles.scrollContainer}>
        <View style={styles.memberContainer}>
          {displayTasks}
        </View>
      </ScrollView>
      {(user.email === project.leaderEmail)
        ? (
          <TouchableHighlight style={styles.button} onPress={() => setModalVisible(true)} underlayColor="#33E86F">
            <Text style={{ textAlign: 'center', fontSize: 20 }}>Create Tasks</Text>
          </TouchableHighlight>
        )
        : <View />}
      <TouchableHighlight style={styles.backButton} onPress={() => setManagingTask(false)} underlayColor="#33E86F">
        <Text style={{ textAlign: 'center', fontSize: 20 }}>Return</Text>
      </TouchableHighlight>
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flexDirection: 'column',
    alignItems: 'center',
    paddingTop: 25,
  },
  projectTitle: {
    fontSize: 40,
    fontWeight: '600',
    textDecorationLine: 'underline',
  },
  scrollContainer: {
    width: 350,
    height: 400,
    backgroundColor: '#B5E48C',
    borderRadius: 10,
  },
  centeredView: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    marginTop: 20,
  },
  addMemberContainer: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  addMember: {
    backgroundColor: 'green',
    width: 40,
    height: 40,
    marginLeft: 20,
    borderRadius: 50,
    alignContent: 'center',
    justifyContent: 'center',
  },
  taskTitle: {
    width: 100,
    fontSize: 14,
  },
  status: {
    width: 75,
    fontSize: 14,
  },
  memberContainer: {
    justifyContent: 'space-between',
    backgroundColor: '#B5E48C',
    width: 350,
    paddingVertical: 30,
    paddingHorizontal: 20,
    borderRadius: 10,
    marginVertical: 30,
  },
  task: {
    flexDirection: 'row',
    backgroundColor: '#6A9B72',
    justifyContent: 'space-between',
    alignItems: 'center',
    padding: 5,
    borderRadius: 10,
    marginBottom: 5,
  },
  input: {
    backgroundColor: 'white',
    borderColor: 'gray',
    width: 250,
    borderWidth: 1,
    borderRadius: 10,
    padding: 10,
    marginVertical: 5,
  },
  button: {
    backgroundColor: '#52B69A',
    borderRadius: 10,
    paddingVertical: 10,
    width: 150,
    justifyContent: 'center',
    alignItems: 'center',
    margin: 10,
    paddingHorizontal: 5,
  },
  backButton: {
    backgroundColor: '#52B69A',
    borderRadius: 10,
    paddingVertical: 10,
    width: 150,
    justifyContent: 'center',
    alignItems: 'center',
    margin: 10,
    paddingHorizontal: 5,
  },
  viewButton: {
    backgroundColor: '#52B69A',
    borderRadius: 10,
    width: 100,
    justifyContent: 'center',
    alignItems: 'center',
    padding: 5,
  },
  cancelButton: {
    backgroundColor: 'red',
    borderRadius: 10,
    paddingVertical: 10,
    width: 150,
    justifyContent: 'center',
    alignItems: 'center',
    margin: 10,
    paddingHorizontal: 5,
  },
});
