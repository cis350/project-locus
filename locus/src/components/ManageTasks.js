/* eslint-disable no-alert */
/* eslint-disable no-unused-expressions */
/* eslint-disable jsx-a11y/label-has-associated-control */
/* eslint-disable no-underscore-dangle */
import React, { useState, useEffect } from 'react';
import {
  Button,
  Row,
  Col,
  Modal,
} from 'react-bootstrap';
import {
  createTask,
  getTasksForProject,
  updateStatusForCurrTask,
  deleteTask,
} from '../modules/api';

const ManageTasks = function ManageTasksComponent({
  project,
  club,
  email,
  role,
}) {
  const [modalClickCreate, setModalClickCreate] = useState(false);
  const [modalClickTask, setModalClickTask] = useState(false);
  const [taskName, setTaskName] = useState('');
  const [targetEmail, setTargetName] = useState('');
  const [allTasks, setAllTasks] = useState([]);
  const [taskAdded, setTaskAdded] = useState(false);
  const [currTask, setCurrTask] = useState('');
  const [currTaskId, setCurrTaskId] = useState('');
  const [currStatus, setCurrStatus] = useState('');
  const [updatedStatus, setUpdatedStatus] = useState('');

  useEffect(() => {
    getTasksForProject(project, club, email).then((res) => {
      if (res.status === 200) {
        setAllTasks(res.jsonContent);
      }
    });
  }, [taskAdded]);

  const openCreateModal = () => {
    setModalClickCreate(true);
  };

  const closeCreateModal = () => {
    setModalClickCreate(false);
  };

  const openTaskModal = (task, id, status) => {
    setModalClickTask(true);
    setCurrTask(task);
    setCurrTaskId(id);
    setCurrStatus(status);
  };

  const closeTaskModal = () => {
    setModalClickTask(false);
  };

  const updateTaskName = ((e) => {
    setTaskName(e.target.value);
  });

  const updateTargetEmail = ((e) => {
    setTargetName(e.target.value);
  });

  const addTask = (() => {
    createTask(project, club, taskName, email, targetEmail, 'incomplete').then((res) => {
      if (res.status === 201) {
        setTaskAdded(!taskAdded);
      } else {
        alert('Error!');
      }
    });
  });

  const createModal = (() => (
    <Modal show={modalClickCreate} onHide={() => closeCreateModal()}>
      <Modal.Header closeButton>
        <Modal.Title>
          Create task for
          {' '}
          {project}
        </Modal.Title>
      </Modal.Header>
      <Modal.Body>
        <input className="form-control" type="text" placeholder="Type in the task name" onChange={(e) => updateTaskName(e)} />
        <input className="form-control" type="text" placeholder="Type in the the recipient email" onChange={(e) => updateTargetEmail(e)} />
      </Modal.Body>
      <Modal.Footer>
        <Button className="btn btn-secondary mt-2" onClick={() => closeCreateModal()}>
          Close
        </Button>
        <Button className="btn btn-primary mt-2" onClick={() => { addTask(); closeCreateModal(); }}>
          Create
        </Button>
      </Modal.Footer>
    </Modal>
  ));

  const updateStatus = (() => {
    setUpdatedStatus('');
  });

  const updateStatusState = ((e) => {
    setUpdatedStatus(e.target.value);
  });

  const updateStatusForTask = (() => {
    updateStatusForCurrTask(currTaskId, club, project, email, updatedStatus).then((res) => {
      if (res.status === 200) {
        setTaskAdded(!taskAdded);
      }
    });
  });

  const deleteCurrTask = (() => {
    deleteTask(currTaskId, club, project, email).then((res) => {
      if (res.status === 200) {
        setTaskAdded(!taskAdded);
      }
    });
  });
  const taskModal = (() => (
    <Modal show={modalClickTask} onHide={() => closeTaskModal()}>
      <Modal.Header closeButton>
        <Modal.Title>
          Update status for
          {' '}
          {currTask}
        </Modal.Title>
      </Modal.Header>
      <Modal.Body>
        Current status:
        {' '}
        {currStatus}
        <div className="form-check">
          <input className="form-check-input" type="radio" name="inlineRadioOptions" id="radio1" value="incomplete" onChange={(e) => updateStatusState(e)} />
          <label className="form-check-label" htmlFor="inlineRadio1">Incomplete</label>
        </div>
        <div className="form-check">
          <input className="form-check-input" type="radio" name="inlineRadioOptions" id="radio2" value="done" onChange={(e) => updateStatusState(e)} />
          <label className="form-check-label" htmlFor="inlineRadio2">Done</label>
        </div>
        <div className="form-check">
          <input className="form-check-input" type="radio" name="inlineRadioOptions" id="radio2" value="need help" onChange={(e) => updateStatusState(e)} />
          <label className="form-check-label" htmlFor="inlineRadio2">Need help</label>
        </div>
      </Modal.Body>
      <Modal.Footer>
        <Button className="btn btn-secondary mt-2" onClick={() => closeTaskModal()}>
          Close
        </Button>
        <Button className="btn btn-primary mt-2" onClick={() => { updateStatus(); updatedStatus !== '' ? updateStatusForTask() : alert('please select an option'); closeTaskModal(); }}>
          Update
        </Button>
        <Button className="btn btn-danger mt-2" onClick={() => { closeTaskModal(); deleteCurrTask(); }}>
          Delete Task
        </Button>
      </Modal.Footer>
    </Modal>
  ));

  const displayProjects = (() => (
    allTasks.map((data) => (
      <div className="club-item" key={data._id}>
        {/* referenced https://www.delftstack.com/howto/javascript/javascript-remove-spaces */}
        <Button className="club-button" onClick={() => openTaskModal(data.taskName, data._id, data.status)}>
          <Row>
            <Col className="d-flex justify-content-center">
              {data.taskName}
              {' '}
              (Assigned to:
              {' '}
              {data.assignedTo}
              {' '}
              Status:
              {' '}
              {data.status}
              )
            </Col>
          </Row>
        </Button>
      </div>
    ))
  ));

  return (
    <div className="container" style={{ position: 'relative', padding: '20px' }}>
      <div className="club-item">
        {/* referenced https://www.delftstack.com/howto/javascript/javascript-remove-spaces */}
        {role === 'master' && (
        <Button className="club-button" onClick={() => openCreateModal()}>
          <Row>
            <Col className="d-flex justify-content-center">
              Create new task
            </Col>
          </Row>
        </Button>
        )}
      </div>
      {allTasks.length !== 0 && displayProjects()}
      {modalClickTask && taskModal()}
      {modalClickCreate && createModal()}
    </div>
  );
};

export default ManageTasks;
