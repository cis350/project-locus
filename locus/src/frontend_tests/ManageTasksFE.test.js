/**
* @jest-environment jsdom
*/
import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import ManageTasks from '../components/ManageTasks';

const axios = require('axios');
const MockAdapter = require('axios-mock-adapter');

// mock useNavigate: https://github.com/remix-run/react-router/issues/7811
const mockedNavigator = jest.fn();
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockedNavigator,
}));

let mock;
beforeAll(() => {
  mock = new MockAdapter(axios);
});
afterEach(() => {
  mock.reset();
});

test('manage tasks snapshot test', () => {
  const component = renderer.create(<ManageTasks />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
