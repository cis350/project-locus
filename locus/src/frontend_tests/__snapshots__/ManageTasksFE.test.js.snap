// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`manage tasks snapshot test 1`] = `
<div
  className="container"
  style={
    Object {
      "padding": "20px",
      "position": "relative",
    }
  }
>
  <div
    className="club-item"
  >
    <button
      className="club-button btn btn-primary"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      <div
        className="row"
      >
        <div
          className="d-flex justify-content-center col"
        >
          Create new task
        </div>
      </div>
    </button>
  </div>
</div>
`;
